pdu.c

```diff
@@ -525,55 +535,157 @@ static int pdu_parse_timestamp(char ** pdu, size_t * length)
 	return -EINVAL;
 }
 
-#/* TODO: remove / TODO: append 8 bit */
-static int check_encoding(const char* msg, unsigned length)
+#define CSMS_GSM7_MAX_LEN 153
+#define SMS_GSM7_MAX_LEN 160
+#define CSMS_UCS2_MAX_LEN 66
+#define SMS_UCS2_MAX_LEN 70
+#define PDU_LENGTH 176
+
+#define USE_BOM 0
+
+EXPORT_DEF int pdu_build_mult(int (*cb)(const char *buf, unsigned len, void *s), const char* sca, const char* dst, const char* msg, unsigned valid_minutes, int srr, unsigned csmsref, void *s)
 {
-	str_encoding_t possible_enc = get_encoding(RECODE_ENCODE, msg, length);
-	if(possible_enc == STR_ENCODING_7BIT_HEX_PAD_0)
-		return PDU_DCS_ALPHABET_7BIT;
-	return PDU_DCS_ALPHABET_UCS2;
+	unsigned length = 2048;
+	char buffer[length];
+	unsigned msg_len, utf16_len;
+
+	msg_len = strlen(msg);
+	utf16_len = msg_len * 2 + 4;
+// 	uint16_t *msg_utf16 = ast_malloc(utf16_len);
+	uint16_t msg_utf16[utf16_len]; // TODO: heap
+	utf16_len = convert_string(msg, msg_len, (char*)msg_utf16, utf16_len, "UTF-8", 
+#if USE_BOM == 1
+		"UTF-16"
+#else
+		"UTF-16BE"
+#endif
+	) / 2;
+// 	uint16_t *msg_gsm7 = ast_malloc(utf16_len * 2);
+	uint16_t msg_gsm7[utf16_len * 2];
+
+	// TODO: Check for other tables
+	int is_gsm7 = 1;
+	unsigned gsm7_len = 0;
+	const uint8_t *escenc = get_char_gsm7_encoding(0x1B00);
+	for (unsigned i = USE_BOM /* 0 is BOM */; i < utf16_len; ++i) {
+
+// 	printf("%d\n", msg_utf16[i]);
+		const uint8_t *enc = get_char_gsm7_encoding(msg_utf16[i]);
+		uint8_t c = enc[0];
+// 		printf("Enc: %d %s\n", c, (c > 127 ? LUT_GSM7_SS : LUT_GSM7_LS)[0][c % 128]);
+// 	printf("%d\n", c);
+		if (c == GSM7_INVALID) {
+// 			printf("%d is not gsm7 @%d %d\n", msg_utf16[i], i, USE_BOM);
+			is_gsm7 = 0;
+			break;
+		}
+		if (c > 127) {
+			gsm7_len += 2;
+// 			printf("%d\n\n", escenc[0]);
+			msg_gsm7[i - USE_BOM] = (escenc[0] << 8) | (c - 128);
+		} else {
+			++gsm7_len;
+			msg_gsm7[i - USE_BOM] = c;
+		}
+	}
+	msg_gsm7[gsm7_len] = '\0';
+
+	unsigned split = 0;
+	unsigned cnt = 0, cur = 1, off = 0;
+	int res;
+// 	printf("IS GSM7: %d len: %d\n", is_gsm7, utf16_len);
```

Comment:
I'm not that fond of variable declarations mid-function.

It generally means your function grew too big, and you should use an auxiliary function instead.

(It's also a reason why you're adding -std=gnu99, but I guess that's not a problem for most people.)